pkg/ebpf/perf/event.go

```diff
@@ -0,0 +1,213 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux_bpf
+
+// Package perf implements types related to eBPF and the perf subsystem, like perf buffers and ring buffers.
+package perf
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/features"
+	"github.com/cilium/ebpf/perf"
+	"github.com/cilium/ebpf/ringbuf"
+
+	ebpfTelemetry "github.com/DataDog/datadog-agent/pkg/ebpf/telemetry"
+	ddsync "github.com/DataDog/datadog-agent/pkg/util/sync"
+)
+
+var perfPool = ddsync.NewDefaultTypedPool[perf.Record]()
+var ringbufPool = ddsync.NewDefaultTypedPool[ringbuf.Record]()
+
+// Flushable is an interface for objects that support flushing
+type Flushable interface {
+	Flush()
+}
+
+// EventHandler abstracts consuming data from a perf buffer or ring buffer (depending on availability and options).
+// It handles upgrading maps from a ring buffer is desired, and unmarshalling into the desired data type.
+type EventHandler struct {
+	f    Flushable
+	opts EventHandlerOptions
+}
+
+// EventHandlerOptions are the options controlling the EventHandler.
+// MapName and Handler are required options.
+type EventHandlerOptions struct {
+	MapName string
+	Handler func([]byte)
+
+	TelemetryEnabled  bool
+	UseRingBuffer     bool
+	UpgradePerfBuffer bool
+
+	PerfOptions    PerfBufferOptions
+	RingBufOptions RingBufferOptions
+}
+
+// PerfBufferOptions are options specifically for perf buffers
+//
+//nolint:revive
+type PerfBufferOptions struct {
+	BufferSize int
+
+	// Watermark - The reader will start processing samples once their sizes in the perf ring buffer
```

Comment:
perf ring buffer?

Reply:
Technically a perf buffer is implemented with a ring buffer. I can see how this is confusing here, so I will reword.